jax/interpreters/partial_eval.py

```diff
@@ -1574,3 +1597,74 @@ def partial_eval_to_jaxpr_dynamic(fun: lu.WrappedFun, in_pvals: Sequence[Partial
   # TODO(mattjj): alias to trace_to_jaxpr after revising custom_derivatives.py
   with core.new_main(core.EvalTrace, dynamic=True) as _:  # type: ignore
     return trace_to_jaxpr(fun, in_pvals)
+
+
+AvalId = int
+TracerId = int
+def _avals_to_tracers(
+    trace: DynamicJaxprTrace, in_avals: Sequence[AbstractValue]
+  )  -> Sequence[Tracer]:
+  # Create input Tracers given input abstract values, handling nesting: each
+  # element `a` of `in_avals` can have abstract values in its shape, which must
+  # occurr to the left of `a`.
+  env: Dict[AvalId, Tracer] = {}
+  in_tracers: List[Tracer] = []
+  for a in in_avals:
+    t = env[id(a)] = trace.new_arg(_subs_tracers_in_aval(env, a))
+    in_tracers.append(t)
+  return in_tracers
+
+def _subs_tracers_in_aval(
+    env: Dict[AvalId, Tracer], a: AbstractValue
+  ) -> AbstractValue:
+  # Substitute Tracers into a given AbstractValue using the given environment.
+  # That is, the input is an AbstractValue possibly containing AbstractValues,
+  # and the output is an aval possibly containing tracers.
+  if (isinstance(a, DShapedArray) and
+      any(isinstance(d, AbstractValue) for d in a.shape)):
+    shape = [env[id(d)] if isinstance(d, AbstractValue) else d for d in a.shape]
+    return a.update(shape=tuple(shape))
+  return a
+
+def _tracers_to_avals(
+    env: Dict[TracerId, AbstractValue], tracers: Sequence[Tracer]
+  ) -> List[AbstractValue]:
+  # Replace Tracers with corresponding abstract values, handling Tracers in
+  # shapes and ensuring each Tracer is mapped to a unique AbstractValue object.
+  avals: List[AbstractValue] = []
+  for t in tracers:
+    aval = env.get(id(t))
+    if aval is None:
+      aval = env[id(t)] = _subs_avals_in_aval(env, t.aval)
+    avals.append(aval)
+  return avals
+
+def _subs_avals_in_aval(
+    env: Dict[TracerId, AbstractValue], a: AbstractValue
+  ) -> AbstractValue:
+  # Substitute AbstractValues into given AbstractValue using given environment.
+  # That is, the input is an AbstractValue possibly containing Tracers and the
+  # output is an AbstractValue possibly containing AbstractValues.
+  if (isinstance(a, DShapedArray) and
+      any(isinstance(d, Tracer) for d in a.shape)):
+    shape = [env.setdefault(id(d), d.aval) if isinstance(d, Tracer) else d
+             for d in a.shape]
+    return a.update(shape=tuple(shape))
+  else:
+    return a
+
+def _get_tracers_only_in_shapes(in_tracers: Sequence[Tracer]) -> List[Tracer]:
+  # In DynamicJaxprTrace.process_call (e.g. for handling jit-of-jit) we want to
+  # extract Tracers from the shapes of arguments so as to elaborate them into
+  # explicit inputs to the call that appears in the jaxpr. Some of these Tracers
+  # may already appear as explicit inputs, so we only need to get those present
+  # exclusively in shapes.
```

Comment:
In order to get only the tracers that are exclusively in shapes, I think that you need to add the `in_tracers` to `seen` below.

Reply:
Thanks for the catch! I'll add a test for this too.

Reply:
Done!